cli/azd/pkg/tools/github/github.go

```diff
@@ -323,3 +377,186 @@ func isEnvVarTokenSource(message string) bool {
 
 	return false
 }
+
+func extractFromZip(src, dst string) (string, error) {
+	zipReader, err := zip.OpenReader(src)
+	if err != nil {
+		return "", err
+	}
+
+	defer zipReader.Close()
+
+	var extractedAt string
+	for _, file := range zipReader.File {
+		if !file.FileInfo().IsDir() && strings.Contains(file.Name, "gh") {
+			fileReader, err := file.Open()
+			if err != nil {
+				return extractedAt, err
+			}
+			fileNameParts := strings.Split(file.Name, "/")
+			fileName := fileNameParts[len(fileNameParts)-1]
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, fileReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+	return extractedAt, nil
+}
+func extractFromTar(src, dst string) (string, error) {
+	gzFile, err := os.Open(src)
+	if err != nil {
+		return "", err
+	}
+	defer gzFile.Close()
+
+	gzReader, err := gzip.NewReader(gzFile)
+	if err != nil {
+		return "", err
+	}
+	defer gzReader.Close()
+
+	var extractedAt string
+	// tarReader doesn't need to be closed as it is closed by the gz reader
+	tarReader := tar.NewReader(gzReader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find gh cli within tar file: %w", err)
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "gh" {
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+
+	return extractedAt, nil
+}
+
+// extractGhCli gets the Github cli from either a zip or a tar.gz
+func extractGhCli(src, dst string) (string, error) {
+	if strings.Contains(src, ".zip") {
```

Comment:
I wonder if `strings.HasSuffix` would give a clearer indication that we were trying to check the extension.

Reply:
I am using `os.CreateTemp(path, name)`. So, for a name like `gh_2.22.1_linux_amd64.tar.gz`, the resulting file is something like `gh_2.22.1_linux_amd64.tar.gz1694433216`

I copied the temp file strategy from how we get `bicep`.  
My understanding is that we use that temp-name while we read and copy all bytes form the network into it (so it does not gives the impression that the file is ready yet).

I guess I can rename the temp file as soon as the download completes and extract from the renamed file, so the sufix is the extension.